sql/src/main/java/io/crate/analyze/MaybeAliasedStatement.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Licensed to Crate under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.analyze;
+
+import io.crate.analyze.relations.AliasedAnalyzedRelation;
+import io.crate.analyze.relations.AnalyzedRelation;
+import io.crate.expression.symbol.Field;
+import io.crate.expression.symbol.FieldReplacer;
+import io.crate.expression.symbol.Symbol;
+
+import java.util.function.Function;
+
+public class MaybeAliasedStatement {
+
+    public static MaybeAliasedStatement analyze(AnalyzedRelation relation) {
+        if (relation instanceof AliasedAnalyzedRelation) {
+            return new MaybeAliasedStatement((AliasedAnalyzedRelation) relation);
+        }
+        return new NoopAliasedStatement(relation);
+    }
+
+    private final AliasedAnalyzedRelation relation;
+    private final Function<? super Field, ? extends Symbol> fieldMapper;
+
+    private MaybeAliasedStatement(AliasedAnalyzedRelation relation) {
+        this.relation = relation;
+        this.fieldMapper = f -> {
+            if (f.relation().equals(relation)) {
+                return f.pointer();
+            }
+            return f;
+        };
+    }
+
+    public AnalyzedRelation relation() {
+        return relation.relation();
+    }
+
+    public Symbol mapFields(Symbol symbol) {
+        return FieldReplacer.replaceFields(symbol, fieldMapper);
+    }
+
+    private static class NoopAliasedStatement extends MaybeAliasedStatement {
```

Comment:
Maybe personal preference, but this sub-classing could be replaced by adding a `Function<Symbol, Symbol>` as argument to `MaybeAliasedStatement`.

Reply:
I think it's more readable as it is now

Reply:
Another function as a parameter would be required for the `relation()` getter.
I'm keeping it as it is right now if ok.

Reply:
Fair enough, this is what I had in mind btw:


```
diff --git a/sql/src/main/java/io/crate/analyze/MaybeAliasedStatement.java b/sql/src/main/java/io/crate/analyze/MaybeAliasedStatement.java
index a19e0b71f..defc7d1a4 100644
--- a/sql/src/main/java/io/crate/analyze/MaybeAliasedStatement.java
+++ b/sql/src/main/java/io/crate/analyze/MaybeAliasedStatement.java
@@ -25,7 +25,6 @@ package io.crate.analyze;
 import io.crate.analyze.relations.AliasedAnalyzedRelation;
 import io.crate.analyze.relations.AnalyzedRelation;
 import io.crate.analyze.relations.FieldResolver;
-import io.crate.expression.symbol.Field;
 import io.crate.expression.symbol.FieldReplacer;
 import io.crate.expression.symbol.Symbol;
 
@@ -41,49 +40,29 @@ public class MaybeAliasedStatement {
 
     public static MaybeAliasedStatement analyze(AnalyzedRelation relation) {
         if (relation instanceof AliasedAnalyzedRelation) {
-            return new MaybeAliasedStatement((AliasedAnalyzedRelation) relation);
+            AliasedAnalyzedRelation aliasedAnalyzedRelation = (AliasedAnalyzedRelation) relation;
+            return new MaybeAliasedStatement(
+                aliasedAnalyzedRelation.relation(),
+                FieldReplacer.bind( f -> f.relation().equals(aliasedAnalyzedRelation) ? f.pointer() : f)
+            );
+        } else {
+            return new MaybeAliasedStatement(relation, x -> x);
         }
-        return new NoopAliasedStatement(relation);
     }
 
-    private final AliasedAnalyzedRelation relation;
-    private final Function<? super Field, ? extends Symbol> fieldMapper;
+    private final AnalyzedRelation nonAliasRelation;
+    private final Function<? super Symbol, ? extends Symbol> mapper;
 
-    private MaybeAliasedStatement(AliasedAnalyzedRelation relation) {
-        this.relation = relation;
-        this.fieldMapper = f -> {
-            if (f.relation().equals(relation)) {
-                return f.pointer();
-            }
-            return f;
-        };
+    private MaybeAliasedStatement(AnalyzedRelation nonAliasRelation, Function<? super Symbol, ? extends Symbol> mapper) {
+        this.nonAliasRelation = nonAliasRelation;
+        this.mapper = mapper;
     }
 
-    public AnalyzedRelation relation() {
-        return relation.relation();
+    public AnalyzedRelation nonAliasRelation() {
+        return nonAliasRelation;
     }
 
     public Symbol maybeMapFields(Symbol symbol) {
-        return FieldReplacer.replaceFields(symbol, fieldMapper);
-    }
-
-    private static class NoopAliasedStatement extends MaybeAliasedStatement {
-
-        private final AnalyzedRelation relation;
-
-        NoopAliasedStatement(AnalyzedRelation relation) {
-            super(null);
-            this.relation = relation;
-        }
-
-        @Override
-        public AnalyzedRelation relation() {
-            return relation;
-        }
-
-        @Override
-        public Symbol maybeMapFields(Symbol symbol) {
-            return symbol;
-        }
+        return mapper.apply(symbol);
     }
 }
``` 

Might want to take over the rename from `relation` to `nonAliasedRelation` anyway as I think it clarifies that the relation that is returned isn't the aliased one.